### PR TITLE
Ticket3894 procserv launcher

### PR DESCRIPTION
**Pull request by Alistair-McGann-Tessella**
See https://github.com/ISISComputingGroup/IBEX/issues/3894

Finishes off the procserv launcher so IOCs can be booted through procserv in the ioc test framework

### CODE REVIEW

File: utils/ioc_launcher.py [hunk 1]
```diff
@@ -202,45 +209,103 @@ def open(self):
...
-                           ' --logport={:d}'.format(self.logport), ' --chdir="{}"'.format(cygwin_dir),
-                           ' {:d}'.format(self.port), ' {}'.format(comspec), ' /c', ' runIOC.bat', ' st.cmd']
+        self.ioc_run_command = ["{}\\cygwin_bin\\procServ.exe".format(self.ICPTOOLS), ' --logstamp',
+                                ' --logfile="{}"'.format(self.to_cygwin_address(self._log_filename())),
+                                ' --timefmt="%Y-%m-%d %H:%M:%S"',
+                                ' --restrict', ' --ignore="^D^C"', ' --autorestart', ' --wait',
+                                ' --name={}'.format(self._device.upper()),
+                                ' --pidfile="/cygdrive/c/windows/temp/EPICS_{}.pid"'.format(self._device),
+                                ' --logport={:d}'.format(self.logport), ' --chdir="{}"'.format(cygwin_dir),
+                                ' {:d}'.format(self.port), ' {}'.format(comspec), ' /c', ' runIOC.bat', ' st.cmd']
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply utils/ioc_launcher.py]: 👍 This also helps clear up the argument comparison later on

File: utils/ioc_launcher.py [hunk 2]
```diff
@@ -202,45 +209,103 @@ def open(self):
...
+        if self.telnet is not None:
+            self.telnet.close()
+
+        cmdline_arguments = [args.strip().replace('"', '') for args in self.ioc_run_command]
+
+        for process in psutil.process_iter(attrs=['pid', 'name']):
+            if process.info['name'] == 'procServ.exe':
+                if all([args in process.cmdline() for args in cmdline_arguments]):
+                    # Command line arguments match, kill procServ instance
+                    os.kill(process.pid, -1)
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply utils/ioc_launcher.py]: I couldn't find a constant in the signals package which is `-1` (kill), so I used `SIGTERM` instead which is similar but can be blocked.

File: utils/ioc_launcher.py [hunk 3]
```diff
@@ -134,10 +137,15 @@ def __init__(self, ioc, test_mode, var_dir):
...
 
+        self.ioc_run_command = None
         self._process = None
         self.log_file_manager = None
         self._ca = None
         self.macros = None
+        self.telnet = None
+        self.autorestart = True
+
+        atexit.register(self.close)
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply utils/ioc_launcher.py]: A context manager's `__exit__` gets run on a control-C event (and any other exception).

File: utils/ioc_launcher.py [hunk 4]
```diff
@@ -202,45 +209,103 @@ def open(self):
...
+        if "Welcome to procServ" not in init_output:
+            raise OSError("Cannot connect to procServ over telnet")
+
+    def start_ioc(self):
+        """
+        Sends the start/restart IOC command to procserv. (^X)
+
+        """
+        start_command = "\x18"
+        self.telnet.write(start_command + "\n")
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply utils/ioc_launcher.py]: We should stick as far as possible to the conventions at https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Python-conventions#use-format-to-construct-strings-from-values . Having said that, you're welcome to petition to change our style guidelines (usually they get changed at a code chat)!

**Pull request merged**